src/core/Auth.cpp

```diff
@@ -0,0 +1,151 @@
+    struct pam_response* pam_reply         = (struct pam_response*)calloc(num_msg, sizeof(struct pam_response));
+
+    for (int i = 0; i < num_msg; ++i) {
+        switch (msg[i]->msg_style) {
+            case PAM_PROMPT_ECHO_OFF:
+            case PAM_PROMPT_ECHO_ON: {
+                g_pAuth->setPrompt(msg[i]->msg);
+                // Some pam configurations ask for the password twice for whatever reason (Fedora su for example)
+                // When the prompt is the same as the last one, I guess our answer can be the same.
+                Debug::log(LOG, "PAM_PROMPT: {}", msg[i]->msg);
+                if (VERIFICATIONSTATE->prompt != VERIFICATIONSTATE->lastPrompt)
+                    g_pAuth->waitForInput();
+
+                if (g_pHyprlock->m_bTerminate)
+                    return PAM_CONV_ERR;
+
+                pam_reply[i].resp = strdup(VERIFICATIONSTATE->input.c_str());
+                break;
+            }
+            case PAM_ERROR_MSG: Debug::log(ERR, "PAM: {}", msg[i]->msg); break;
+            case PAM_TEXT_INFO: Debug::log(LOG, "PAM: {}", msg[i]->msg); break;
+        }
+    }
+
+    VERIFICATIONSTATE->waitingForPamAuth = true;
+
+    *resp = pam_reply;
+    return PAM_SUCCESS;
+}
+
+static void passwordCheckTimerCallback(std::shared_ptr<CTimer> self, void* data) {
+    g_pHyprlock->onPasswordCheckTimer();
+}
+
+void CAuth::start() {
+    m_authThread = std::thread([this]() {
+        static auto* const PPAMMODULE = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:pam_module"));
+        while (!g_pHyprlock->m_bTerminate) {
+            bool success = auth(*PPAMMODULE);
+            g_pHyprlock->addTimer(std::chrono::milliseconds(1), passwordCheckTimerCallback, nullptr);
+            if (success)
+                break;
+
+            std::this_thread::sleep_for(std::chrono::seconds(1)); // TODO: Make sure that the callback succeeded
+            resetConversation();
+        }
+    });
+
+    m_authThread.detach();
+}
+
+bool CAuth::auth(std::string pam_module) {
+    const pam_conv localConv   = {conv, (void*)&m_sConversationState};
+    pam_handle_t*  handle      = NULL;
+    auto           uidPassword = getpwuid(getuid());
+
+    int            ret = pam_start(pam_module.c_str(), uidPassword->pw_name, &localConv, &handle);
+
+    if (ret != PAM_SUCCESS) {
+        m_sConversationState.success    = false;
+        m_sConversationState.failReason = "pam_start failed";
+        Debug::log(ERR, "auth: pam_start failed for {}", pam_module);
+        return false;
+    }
+
+    ret = pam_authenticate(handle, 0);
+
+    m_sConversationState.waitingForPamAuth = false;
+
+    if (ret != PAM_SUCCESS) {
+        m_sConversationState.success    = false;
+        m_sConversationState.failReason = ret == PAM_AUTH_ERR ? "Authentication failed" : "pam_authenticate failed";
+        Debug::log(ERR, "auth: {} for {}", m_sConversationState.failReason, pam_module);
+        return false;
+    }
+
```

Comment:
I don't mind at all. I feel like this taught me a lot actually since this implementation is a lot better, and truthfully I've also been thinking that a socket was overkill after implementing it. On that note: I think it might be relevant to call [pam_acct_mgmt](https://man7.org/linux/man-pages/man3/pam_acct_mgmt.3.html) here and maybe also [pam_setcred(pamh, PAM_REFRESH_CRED)](https://linux.die.net/man/3/pam_setcred) after that.

Reply:
Cool that you learned smth. I learned that pam is a bit of a mess ^^.

I don't think we want `pam_acct_mgmt`.
That is because the user that locks the screen has to unlock it. It is not the job of the lockscreen to check if that user's account is valid.

Similiar thing with `pam_set_cred`. Irrelevant for a lockscreen.

Greetd has to probably care about that but not us luckily :)